### PR TITLE
Compile RBIs for foreign constants

### PR DESCRIPTION
**Pull request by egiurleo**
### Motivation
Part of the work for #890. Heavily based on work I did with @paracycle on [this branch](https://github.com/Shopify/tapioca/compare/uk-eg-mixin-fix).

Currently, Tapioca only generates RBI for constants that are defined in the gem that's currently being compiled. This is the correct behavior in most cases, except that in Ruby, a gem can add functionality to any constant, even one that it doesn't define.

Let's say there's a gem `foo` that defines the class `Foo`.
Then let's say there's another gem `bar` that defines the module `Bar` and contains the following code:

```
Foo.prepend(Bar)
```

or

```
class Foo
   prepend Bar
end
```

Currently, Tapioca would not recognize that `Foo` had `Bar` mixed in and would generate incomplete RBIs for `Foo`. This is especially true if `Foo` is a constant defined in the Ruby standard library.

We're calling these "foreign constants."

### Implementation
There are a few steps to this implementation:
- Modify the `Mixin` tracker to keep a map between constants and classes that mix them in. The tracker currently keeps a map the other way around, but I think it makes more sense to flip the keys and values in this case (will explain more in next bullet point).
- Introduce a new listener called `ForeignConstants`. This listener takes a constant and uses the `Mixin` tracker to identify all the other constants that have the first constant as a mixin. This will include any constants that are defined in another gem or in the Ruby standard library that have mixed in constants defined in the gem being compiled. All of these constants are then added to the pipeline in a `Gem::ForeignConstantFound` event.
- This event can be processed in a simpler way than the `Gem::ConstantFound` event because there are certain logical branches that don't apply to it and can be skipped entirely. While processing this event, the pipeline knows to skip the `defined_in_gem?` check.
- This will lead to a `ForeignScopeNodeAdded` event being added to the pipeline. The listeners have been modified such that only the `Mixins` listener will accept the `ForeignScopeNodeAdded` event. This way, only mixin RBIs are generated for foreign constants and all other RBIs (method definitions, documentation, etc.) are ignored.

### Tests
I have added multiple pipeline tests and a gem test that fail on main but pass with these changes.

### Tophatting 🎩 

I have tested these changes on Shopify core. It compiles all gem RBIs and results in no type errors!

### CODE REVIEW

File: lib/tapioca/gem/pipeline.rb
Comment (egiurleo) [inline lib/tapioca/gem/pipeline.rb:234]: I'm not very confident about this flag because of the way it's being passed all the way down from the event into the `compile_module` method. I'm curious what the `strict` keyword is supposed to do and if that expresses a similar concept.

File: lib/tapioca/runtime/trackers/mixin.rb
Comment (egiurleo) [inline lib/tapioca/runtime/trackers/mixin.rb:10]: I changed the name of this variable from `mixin_map` to `constants_to_mixin_location`. I had a really hard time remembering what it was supposed to do and I think making it clear what it is mapping between would be helpful. Open to thoughts and name suggestions!

File: lib/tapioca/runtime/trackers/mixin.rb
Comment (egiurleo) [inline lib/tapioca/runtime/trackers/mixin.rb:29]: Changed the name of the variable here for clarity.

File: lib/tapioca/runtime/trackers/mixin.rb
Comment (egiurleo) [inline lib/tapioca/runtime/trackers/mixin.rb:49]: Keeping this hash super simple for right now, but I think we could add the actual locations of the mixins in a follow-up PR.

**Review by paracycle — COMMENTED** [review]
This looks quite complete already. I've added some inline comments.

File: lib/tapioca/gem/events.rb
Comment (paracycle) [inline lib/tapioca/gem/events.rb:45]: What if this subclassed `ConstantFound`? We are basically using the class as a discriminator on how to act, otherwise, all aspects of this class should be the same as `ConstantFound`.

File: lib/tapioca/gem/pipeline.rb
Comment (paracycle) [inline lib/tapioca/gem/pipeline.rb:133]: If `NotOwnedConstantFound` subclasses `ConstantFound`, we don't need this addition.

File: lib/tapioca/gem/pipeline.rb
Comment (paracycle) [inline lib/tapioca/gem/pipeline.rb:151]: If `NotOwnedConstantFound` subclasses `ConstantFound`, we don't need this `T.any`.

**Review by paracycle — COMMENTED** [review]
This looks much much cleaner. I think one more pass will be enough. I am still not in love with the name we have of the concept, so it might be good to get feedback from other folks by turning this into a proper PR.

File: lib/tapioca/gem/pipeline.rb
Comment (Morriar) [inline lib/tapioca/gem/pipeline.rb:167]: If you use `event.constant` you won't need to cast it:

```suggestion
          compile_not_owned_constant(name, event.constant)
```

I think we could refactor a bit the code to not use a local variable.

File: lib/tapioca/gem/pipeline.rb
Comment (paracycle) [inline lib/tapioca/gem/pipeline.rb:169]: Can we turn this into a `case/when` to make it explicit that we are matching on the type of `event`?
```suggestion
        case event
        when Gem::NotOwnedConstantFound
          compile_not_owned_constant(name, event.constant)
        else
          compile_constant(name, event.constant)
        end
```

File: lib/tapioca/gem/listeners/not_owned_constants.rb
Comment (paracycle) [inline lib/tapioca/gem/listeners/not_owned_constants.rb:7]: I am still not a fan of this name, tbh. What do you think of:
- `ForeignConstant` as the name of a constant that we discover that we need to handle but wasn't declared by the current gem.
- `ForeignConstantFound` as the event name

File: lib/tapioca/gem/listeners/not_owned_constants.rb
Comment (paracycle) [inline lib/tapioca/gem/listeners/not_owned_constants.rb:23]: This comment is correct but I think we can improve it slightly. What do you think of this:
```suggestion
          # There are cases where we want to process constants not declared by the current
          # gem, i.e. "foreign constant". These are constants defined in another gem 
          # that this gem is applying a mix-in to. This pattern is especially common 
          # for gems that add behaviour to Ruby standard library classes by mixing in
          # modules to them.
          #
          # The way we identify these "foreign constants" is by asking the mixin
          # tracker which constants have mixed in the current module that we are
          # handling. We add all those constants that we discover to the pipeline 
          # to be processed.
```

File: lib/tapioca/gem/listeners/not_owned_constants.rb
Comment (paracycle) [inline lib/tapioca/gem/listeners/not_owned_constants.rb:17]: I think we can return early if `mixin` is a `Class` instance, since no `Class` can be mixed into something else:
```suggestion
          mixin = event.constant

          return if Class === mixin # Classes can't be mixed into anything
```

File: lib/tapioca/gem/events.rb
Comment (paracycle) [reply lib/tapioca/gem/events.rb:54]: I wonder if we should do the breakdown of the kinds of constants at push time, so that these become 4 separate events instead:
1. `AliasConstantFound`
2. `ValueConstantFound`
3. `ModuleConstantFound`
4. `ForeignModuleConstantFound < ModuleConstantFound`

All we would need to do is to move the logic in `compile_constant` inside `push_constant` and the handling becomes much simpler. WDYT @Morriar?

Comment (Morriar) [reply lib/tapioca/gem/events.rb:54]: Yeah I like it!

Another option to keep the original design is to use a cast:

```rb
class ForeignConstantFound < ConstantFound
  extend T::Sig

  sig { override.returns(Module) }
  def constant
    T.cast(@constant, Module)
  end

  sig { params(symbol: String, constant: Module).void }
  def initialize(symbol, constant)
    super
  end
end
```

I'm good with both approaches 👍

**Comment by egiurleo** [comment]
Spent some time pairing with @Morriar to debug the errors that were occurring when trying to compile gem types on Shopify core. This is what we discovered:
- The errors were occurring because tapoica loads all of Shopify core when compiling gems. This means that when core mixes in certain gems (e.g. `ActiveModel`), the mixin tracker will track those classes from core. The `NotOwnedConstants` listener will then add those classes from core to the pipeline, and tapioca will try to compile them. Obviously this is not desired behavior.
- I think the actual issue is that the mixin tracker is casting too wide of a net. It's tracking even the mixins that happen outside of the gem we're currently compiling. This seems related to the other issue that #890 is trying to address, so it might make sense to try to address that issue first, then revisit this PR.

My next step will be to write a failing test to make sure the issue is captured in a repeatable way.

File: lib/tapioca/gem/listeners/foreign_constants.rb
Comment (egiurleo) [inline lib/tapioca/gem/listeners/foreign_constants.rb:19]: @paracycle Updated the comment here! It is almost identical to what you suggested with a couple of small wording changes.

File: lib/tapioca/gem/listeners/foreign_constants.rb
Comment (egiurleo) [inline lib/tapioca/gem/listeners/foreign_constants.rb:29]: I was experiencing an issue where running `tapioca gem` on Shopify core was generating RBIs for ALL mixins, not just mixins performed in the gem currently being compiled. This is my fix for this issue. The mixin tracker keeps track of the location where this mixin took place, and checks that it occurred within the gem before generating RBIs.

Two points:
- Currently, the locations are an array of file paths, so I took the first one here, which I believe should correspond to the "closest" location where the mixin took place? Let me know if that sounds right for now.
- I'm going to create a follow-up PR that changes the notion of what a mixin location is, based on the work @paracycle did in [this branch](https://github.com/Shopify/tapioca/compare/uk-eg-mixin-fix#) -- basically, taking the most recent "require" location as the correct location. I wanted to save that for another PR because this one is already quite big.

Comment (paracycle) [reply lib/tapioca/gem/listeners/foreign_constants.rb:29]: It is going to change soon, as you said, but up to this point, our thinking was "this is coming from this gem if any of the locations match the gem require path prefix". So we used to have this method that wasn't being used, that we can bring back again for this: https://github.com/Shopify/tapioca/commit/bdf5d45cd74f329fd438dd8a9b54fd7f4205a39d

File: lib/tapioca/gem/listeners/foreign_constants.rb
Comment (egiurleo) [inline lib/tapioca/gem/listeners/foreign_constants.rb:28]: @paracycle Here's the use of that method. As far as I could tell, we were never actually using the `mixin_type` from the tracker, so I've just been storing the constant and locations. Let me know if this looks alright!

Comment (paracycle) [reply lib/tapioca/gem/listeners/foreign_constants.rb:28]: Yes, for this part of the work, this much is enough. 

However, later on we will also have to revert https://github.com/Shopify/tapioca/pull/668 at which point, we will have start deciding, for a given mixin and a given mixin type (`include`, `extend`, `prepend`), is that happening from inside this gem or not. At that point, we will need that logic.

File: tasks/readme.rake
Comment (egiurleo) [inline tasks/readme.rake:8]: This is a patch to help fix CI that will be merged soon (#965). Once it's merged I'll rebase on main.

**Review by paracycle — APPROVED** [review]
Looks great! Thanks for all the work you put into this.

The `should_skip?` mechanism is a little bit kludgy but I think we can refine it afterwards.

**Comment by egiurleo** [comment]
@paracycle Yay! 100% agreed that `should_skip?` is kludgy (great word) but I am totally happy to make it better later.

I'll merge this once #965 or something similar is merged.

File: lib/tapioca/gem/events.rb
Comment (vinistock) [inline lib/tapioca/gem/events.rb:55]: Why do we need to define this method if all it does is invoke `super`?

File: lib/tapioca/gem/listeners/foreign_constants.rb
Comment (vinistock) [inline lib/tapioca/gem/listeners/foreign_constants.rb:28]: Do we need both of these checks? If it's mixed in by a gem, it can't be defined by the application, right?

Comment (paracycle) [reply lib/tapioca/gem/listeners/foreign_constants.rb:28]: We need both of them since our application code is split into a lot gems. But this part is temporary code until the next PR which will have better attribution to code that triggered a specific tracked event.

File: lib/tapioca/gem/listeners/base.rb
Comment (egiurleo) [reply lib/tapioca/gem/listeners/base.rb:49]: I like that! Would you be okay with me doing that in a follow-up PR?

File: lib/tapioca/gem/events.rb
Comment (egiurleo) [reply lib/tapioca/gem/events.rb]: I created an issue to follow up on this: #977

**Review by Morriar — APPROVED** [review]

**Pull request merged**
